coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRecordSerde.java

```diff
@@ -0,0 +1,122 @@
+import java.nio.ByteBuffer;
+
+/**
+ * Serializer/Deserializer for {@link CoordinatorRecord}. The format is defined below:
+ * <pre>
+ *     record_key   = [record_type key_message]
+ *     record_value = [value_version value_message]
+ *
+ *     record_type     : The record type is currently define as the version of the key
+ *                       {@link ApiMessageAndVersion} object.
+ *     key_message     : The serialized message of the key {@link ApiMessageAndVersion} object.
+ *     value_version   : The value version is currently define as the version of the value
+ *                       {@link ApiMessageAndVersion} object.
+ *     value_message   : The serialized message of the value {@link ApiMessageAndVersion} object.
+ * </pre>
+ */
+public abstract class CoordinatorRecordSerde implements Serializer<CoordinatorRecord>, Deserializer<CoordinatorRecord> {
+    @Override
+    public byte[] serializeKey(CoordinatorRecord record) {
+        // Record does not accept a null key.
+        return MessageUtil.toVersionPrefixedBytes(
+            record.key().version(),
+            record.key().message()
+        );
+    }
+
+    @Override
+    public byte[] serializeValue(CoordinatorRecord record) {
+        // Tombstone is represented with a null value.
+        if (record.value() == null) {
+            return null;
+        } else {
+            return MessageUtil.toVersionPrefixedBytes(
+                record.value().version(),
+                record.value().message()
+            );
+        }
+    }
+
+    @Override
+    public CoordinatorRecord deserialize(
+        ByteBuffer keyBuffer,
+        ByteBuffer valueBuffer
+    ) throws RuntimeException {
+        final short recordType = readVersion(keyBuffer, "key");
+        final ApiMessage keyMessage = apiMessageKeyFor(recordType);
+        readMessage(keyMessage, keyBuffer, recordType, "key");
+
+        if (valueBuffer == null) {
+            return new CoordinatorRecord(new ApiMessageAndVersion(keyMessage, recordType), null);
+        }
+
+        final ApiMessage valueMessage = apiMessageValueFor(recordType);
+        final short valueVersion = readVersion(valueBuffer, "value");
+        readMessage(valueMessage, valueBuffer, valueVersion, "value");
+
+        return new CoordinatorRecord(
+            new ApiMessageAndVersion(keyMessage, recordType),
+            new ApiMessageAndVersion(valueMessage, valueVersion)
+        );
+    }
+
+    private short readVersion(ByteBuffer buffer, String name) throws RuntimeException {
+        try {
+            return buffer.getShort();
+        } catch (BufferUnderflowException ex) {
+            throw new RuntimeException(String.format("Could not read version from %s's buffer.", name));
+        }
+    }
+
+    private void readMessage(ApiMessage message, ByteBuffer buffer, short version, String name) throws RuntimeException {
+        try {
+            message.read(new ByteBufferAccessor(buffer), version);
```

Comment:
out of curiosity. Do we have validation of the version range? or that is unnecessary to add such validation for those messages?

Reply:
Yes, that validation code is added as part of auto-generated code.
For each json file - there is an autogenerated java file which has a read method which has that validation.

Reply:
> For each json file - there is an autogenerated java file which has a read method which has that validation.

IIRC, the auto-generated java file does not check the version range. the unknown (newer) message may have new fields and the "old" reader can see error if it is lucky. otherwise, it may get incorrect parsed values.

Reply:
But that will be needed for version changes from upstream.

Reply:
> But that will be needed for version changes from upstream.

yes, the scenario I imagined is there are some "new" messages produced by the new brokers. And then we run the older brokers to load those "new" messages. I assumed it should throw error likes `Unknown xxx Unknown`. WDYT? is it valid? in the old coordinator, it checks version range when loading group and offsets (https://github.com/apache/kafka/blob/4ae0ab38dd3a82cf2c36032014ee74af63251859/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L1183)

Reply:
So all this is existing code - it was initially part of GroupCoordinatorRecordSerde.java. Since share coordinator has extra records unrelated to group coordinator, common code was generalized to CoordinatorRecordSerde class and both group and share coordinator inherit from it.

Since the original code has been working and I did not author it, I will not be able to comment on the design.